internal/functions/compare_versions.go

```diff
@@ -0,0 +1,163 @@
+package functions
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/function"
+)
+
+// Ensure the implementation satisfies the desired interfaces.
+var _ function.Function = CompareVersionsFunction{}
+
+func NewCompareVersionsFunction() function.Function {
+	return &CompareVersionsFunction{}
+}
+
+type CompareVersionsFunction struct{}
+
+func (f CompareVersionsFunction) Metadata(ctx context.Context, req function.MetadataRequest, resp *function.MetadataResponse) {
+	resp.Name = "compare_versions"
+}
+
+func (f CompareVersionsFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
+	resp.Definition = function.Definition{
+		Summary:     "Compare two version strings",
+		Description: "Given two version strings and a comparison operator, returns a boolean result of the comparison.",
+
+		Parameters: []function.Parameter{
+			function.StringParameter{
+				Name:        "version1",
+				Description: "First version string",
+			},
+			function.StringParameter{
+				Name:        "operator",
+				Description: "Comparison operator",
+			},
+			function.StringParameter{
+				Name:        "version2",
+				Description: "Second version string",
+			},
+		},
+		Return: function.BoolReturn{},
+	}
+}
+
+func (f CompareVersionsFunction) Run(ctx context.Context, req function.RunRequest, resp *function.RunResponse) {
+	var version1, version2, operator string
+
+	resp.Error = function.ConcatFuncErrors(req.Arguments.Get(ctx, &version1, &operator, &version2))
+	if resp.Error != nil {
+		return
+	}
+
+	versionResult1 := ParseVersion(version1)
+
+	if versionResult1.Error == "unknown" {
+		resp.Error = function.NewFuncError(fmt.Sprintf("Invalid APIC version provided: %s", version1))
+		return
+	}
+
+	versionResult2 := ParseVersion(version2)
+
+	if versionResult2.Error == "unknown" {
+		resp.Error = function.NewFuncError(fmt.Sprintf("Invalid APIC version provided: %s", version2))
+		return
+	}
+
+	v1 := *versionResult1.Version
+
+	v2 := *versionResult2.Version
+
+	var result bool
+	switch operator {
+	case "==":
+		result = IsVersionEqual(v1, v2)
+	case "!=":
+		result = IsVersionNotEqual(v1, v2)
+	case ">":
+		result = IsVersionGreater(v1, v2)
+	case "<":
+		result = IsVersionLesser(v1, v2)
+	case ">=":
+		result = IsVersionGreaterOrEqual(v1, v2)
+	case "<=":
+		result = IsVersionLessOrEqual(v1, v2)
+	default:
+		resp.Error = function.NewFuncError(fmt.Sprintf("Invalid operator: %s", operator))
+		return
+	}
+
```

Comment:
I think we should create a separate function for this and use it here and in provider_test.go so you don't repeat the switch/case in the test. This will be helped by moving the fct files in provider itself.

Reply:
ok